fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/fluid/FluidApi.java

```diff
@@ -0,0 +1,82 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.transfer.v1.fluid;
+
+import com.google.common.base.Preconditions;
+
+import net.minecraft.fluid.Fluid;
+import net.minecraft.item.Item;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.math.Direction;
+
+import net.fabricmc.fabric.api.lookup.v1.block.BlockApiLookup;
+import net.fabricmc.fabric.api.lookup.v1.block.BlockApiLookupRegistry;
+import net.fabricmc.fabric.api.lookup.v1.item.ItemApiLookup;
+import net.fabricmc.fabric.api.lookup.v1.item.ItemApiLookupRegistry;
+import net.fabricmc.fabric.api.transfer.v1.context.ContainerItemContext;
+import net.fabricmc.fabric.api.transfer.v1.item.ItemPreconditions;
+import net.fabricmc.fabric.api.transfer.v1.storage.Storage;
+import net.fabricmc.fabric.impl.transfer.fluid.EmptyItemsRegistry;
+import net.fabricmc.fabric.impl.transfer.fluid.FluidApiImpl;
+import net.fabricmc.fabric.impl.transfer.fluid.SimpleFluidContainingItem;
+
+public final class FluidApi {
+	public static final BlockApiLookup<Storage<Fluid>, Direction> SIDED =
+			BlockApiLookupRegistry.getLookup(new Identifier("fabric:sided_fluid_api"), Storage.asClass(), Direction.class);
+
+	public static final ItemApiLookup<Storage<Fluid>, ContainerItemContext> ITEM =
+			ItemApiLookupRegistry.getLookup(new Identifier("fabric:fluid_api"), Storage.asClass(), ContainerItemContext.class);
+
+	/**
+	 * Register an item that contains a fluid and can be emptied of it entirely.
+	 * @param fullItem The item that contains the fluid.
+	 * @param emptyItem The emptied item.
+	 * @param fluid The contained fluid. May not be empty.
+	 * @param amount The amount of fluid in the full item. Must be positive.
+	 */
+	// TODO: document conflicts
+	public static void registerFullItem(Item fullItem, Item emptyItem, Fluid fluid, long amount) {
+		ItemPreconditions.notEmpty(fullItem);
+		FluidPreconditions.notEmpty(fluid);
+		Preconditions.checkArgument(amount > 0);
+
+		ITEM.register((key, ctx) -> new SimpleFluidContainingItem(ctx, key, emptyItem, fluid, amount), fullItem);
+	}
+
+	/**
+	 * Register an item that is empty, and may be filled with some fluid entirely.
+	 */
+	// TODO: document params and conflicts
+	// TODO: pick parameter order, probably the same for both methods?
+	public static void registerEmptyItem(Item emptyItem, Item fullItem, Fluid fluid, long amount) {
+		ItemPreconditions.notEmpty(emptyItem);
+		ItemPreconditions.notEmpty(fullItem);
+		FluidPreconditions.notEmpty(fluid);
+		Preconditions.checkArgument(amount > 0);
+
+		EmptyItemsRegistry.registerEmptyItem(emptyItem, fullItem, fluid, amount);
+	}
+
+	// TODO: potion handling
```

Comment:
Not exactly. The plan is to allow any mod to define a mapping `Potion` -> `(Fluid, long)` so that the `Storage<Fluid>` for a full potion can correctly empty potions with a registered fluid mapping. In particular, this would only be used by the API for water bottles which are actually `Potions.Water`, but other mods may use it to register a `Fluid` per `Potion` if they want to.

Reply:
Ah, I see. Sounds useful.